File: packages/domain/src/accessforge_domain/canonical.py
"""RFC8785 (JSON Canonicalization Scheme) serialization and SHA-256 digests.

Every digest-bound guarantee in AccessForge rests on this file: sealed manifests, approval binding,
and the evidence hash chain. A TypeScript implementation in packages/contracts/ts must produce
byte-identical output; shared vectors in packages/contracts/vectors/canonicalization.json are the
arbiter, and neither language owns them.

Two deliberate restrictions, recorded in ADR 0003:

* **Non-integral floats are rejected.** Guaranteeing identical shortest-round-trip float formatting
  across languages is difficult, and CONTRACTS section 4 states that no floating-point value may
  grant authority or decide an outcome. Refusing them is better than a silent digest divergence.
* **Integers outside the JSON safe range are rejected**, per the same section's overflow rule.
"""

from __future__ import annotations

import hashlib
import math
from typing import Any

MAX_SAFE_INTEGER = 2**53 - 1
MIN_SAFE_INTEGER = -(2**53) + 1

# RFC8785 uses the JSON short escapes where they exist and lower-case \u00xx otherwise.
_SHORT_ESCAPES = {
    0x08: "\\b",
    0x09: "\\t",
    0x0A: "\\n",
    0x0C: "\\f",
    0x0D: "\\r",
    0x22: '\\"',
    0x5C: "\\\\",
}


class CanonicalizationError(ValueError):
    """A value cannot be canonicalized deterministically.

    Raised rather than coerced. A canonicalizer that guessed would produce a digest that two
    implementations could disagree about, which is the failure this module exists to prevent.
    """


def _reject_lone_surrogates(value: str) -> None:
    """Refuse text containing an unpaired surrogate code unit.

    A lone surrogate has no UTF-8 encoding. Left unchecked, Python raised a raw
    UnicodeEncodeError from inside digest() while TypeScript silently substituted U+FFFD and
    returned a confident, wrong hash — the more dangerous of the two, and precisely the kind of
    successful-looking wrong answer this product exists to prevent. Both implementations now
    refuse at the same point with the same error.
    """
    index = 0
    length = len(value)
    while index < length:
        code = ord(value[index])
        if 0xD800 <= code <= 0xDBFF:  # high surrogate: must be followed by a low surrogate
            if index + 1 >= length or not (0xDC00 <= ord(value[index + 1]) <= 0xDFFF):
                raise CanonicalizationError(
                    f"unpaired high surrogate U+{code:04X} at index {index}: no UTF-8 encoding "
                    "exists, so no digest over it can be meaningful"
                )
            index += 2
            continue
        if 0xDC00 <= code <= 0xDFFF:  # low surrogate with no preceding high surrogate
            raise CanonicalizationError(
                f"unpaired low surrogate U+{code:04X} at index {index}: no UTF-8 encoding exists"
            )
        index += 1


def _serialize_string(value: str) -> str:
    _reject_lone_surrogates(value)
    out = ['"']
    for char in value:
        code = ord(char)
        escape = _SHORT_ESCAPES.get(code)
        if escape is not None:
            out.append(escape)
        elif code < 0x20:
            out.append(f"\\u{code:04x}")
        else:
            out.append(char)
    out.append('"')
    return "".join(out)


def _serialize_number(value: int | float) -> str:
    if isinstance(value, bool):  # bool is a subclass of int; handled by the caller
        raise CanonicalizationError("bool must be serialized as a literal, not a number")

    if isinstance(value, float):
        if math.isnan(value) or math.isinf(value):
            raise CanonicalizationError(f"{value!r} is not representable in JSON")
        if not value.is_integer():
            raise CanonicalizationError(
                f"refusing to canonicalize the non-integral float {value!r}: cross-language "
                "formatting agreement is not guaranteed, and floating-point values may not "
                "decide an outcome (CONTRACTS section 4)"
            )
        # Integral floats normalize to integers, which also renders -0.0 as 0 exactly as ES6
        # Number::toString does.
        value = int(value)

    if not MIN_SAFE_INTEGER <= value <= MAX_SAFE_INTEGER:
        raise CanonicalizationError(
            f"integer {value} is outside the JSON safe-integer range "
            f"[{MIN_SAFE_INTEGER}, {MAX_SAFE_INTEGER}]; overflow must be rejected"
        )
    return str(value)


def _utf16_sort_key(key: str) -> bytes:
    """Sort key ordering object properties by UTF-16 code unit, as RFC8785 requires.

    Python's natural string ordering compares code points, which differs above the BMP. By code
    point U+FF00 (0xFF00) sorts before U+1F600 (0x1F600); by UTF-16 code unit the order reverses,
    because U+1F600 is the surrogate pair 0xD83D 0xDE00 and 0xD83D < 0xFF00.
    """
    return key.encode("utf-16-be")


def _serialize(value: Any) -> str:
    if value is None:
        return "null"
    if value is True:
        return "true"
    if value is False:
        return "false"
    if isinstance(value, str):
        return _serialize_string(value)
    if isinstance(value, (int, float)):
        return _serialize_number(value)
    if isinstance(value, (list, tuple)):
        return "[" + ",".join(_serialize(item) for item in value) + "]"
    if isinstance(value, dict):
        for key in value:
            if not isinstance(key, str):
                raise CanonicalizationError(
                    f"object keys must be strings, found {type(key).__name__}"
                )
            # Validate before sorting: _utf16_sort_key encodes, and encoding a lone surrogate
            # raises UnicodeEncodeError before serialization would have caught it.
            _reject_lone_surrogates(key)
        items = sorted(value.items(), key=lambda kv: _utf16_sort_key(kv[0]))
        return "{" + ",".join(f"{_serialize_string(k)}:{_serialize(v)}" for k, v in items) + "}"
    raise CanonicalizationError(f"type {type(value).__name__} cannot be canonicalized")


def canonicalize(value: Any) -> str:
    """Return the RFC8785 canonical JSON form of ``value``."""
    return _serialize(value)


def digest(value: Any) -> str:
    """Lower-case hex SHA-256 of the canonical form."""
    return hashlib.sha256(canonicalize(value).encode("utf-8")).hexdigest()

File: packages/contracts/ts/src/canonical.ts
/**
 * RFC8785 (JSON Canonicalization Scheme) serialization and SHA-256 digests.
 *
 * This must agree byte-for-byte with the Python implementation in
 * packages/domain/src/accessforge_domain/canonical.py. The shared vectors in
 * ../../vectors/canonicalization.json are the arbiter; neither language owns them.
 *
 * The same two restrictions apply here, for the same reasons (ADR 0003): non-integral floats and
 * integers outside the JSON safe range are rejected rather than guessed at.
 */

import { createHash } from 'node:crypto';

export const MAX_SAFE = Number.MAX_SAFE_INTEGER; // 2^53 - 1
export const MIN_SAFE = Number.MIN_SAFE_INTEGER;

export class CanonicalizationError extends Error {
  override name = 'CanonicalizationError';
}

const SHORT_ESCAPES = new Map<number, string>([
  [0x08, '\\b'],
  [0x09, '\\t'],
  [0x0a, '\\n'],
  [0x0c, '\\f'],
  [0x0d, '\\r'],
  [0x22, '\\"'],
  [0x5c, '\\\\'],
]);

/**
 * Refuse text containing an unpaired surrogate code unit.
 *
 * A lone surrogate has no UTF-8 encoding. Without this check Node's encoder silently substitutes
 * U+FFFD and `digest()` returns a confident, wrong hash, while the Python implementation raised a
 * raw UnicodeEncodeError. Neither is agreement, and the silent-success side is the more dangerous.
 */
function rejectLoneSurrogates(value: string): void {
  for (let i = 0; i < value.length; i += 1) {
    const code = value.charCodeAt(i);
    if (code >= 0xd800 && code <= 0xdbff) {
      const next = i + 1 < value.length ? value.charCodeAt(i + 1) : -1;
      if (!(next >= 0xdc00 && next <= 0xdfff)) {
        throw new CanonicalizationError(
          `unpaired high surrogate U+${code.toString(16).toUpperCase().padStart(4, '0')} at ` +
            `index ${i}: no UTF-8 encoding exists, so no digest over it can be meaningful`,
        );
      }
      i += 1;
      continue;
    }
    if (code >= 0xdc00 && code <= 0xdfff) {
      throw new CanonicalizationError(
        `unpaired low surrogate U+${code.toString(16).toUpperCase().padStart(4, '0')} at ` +
          `index ${i}: no UTF-8 encoding exists`,
      );
    }
  }
}

function serializeString(value: string): string {
  rejectLoneSurrogates(value);
  let out = '"';
  // Iterating by code unit (not by code point) keeps lone surrogates from being reordered or
  // silently replaced, and matches the ordering rule used for object keys.
  for (let i = 0; i < value.length; i += 1) {
    const code = value.charCodeAt(i);
    const short = SHORT_ESCAPES.get(code);
    if (short !== undefined) {
      out += short;
    } else if (code < 0x20) {
      out += `\\u${code.toString(16).padStart(4, '0')}`;
    } else {
      out += value[i];
    }
  }
  return `${out}"`;
}

function serializeNumber(value: number): string {
  if (!Number.isFinite(value)) {
    throw new CanonicalizationError(`${value} is not representable in JSON`);
  }
  if (!Number.isInteger(value)) {
    throw new CanonicalizationError(
      `refusing to canonicalize the non-integral number ${value}: cross-language formatting ` +
        'agreement is not guaranteed, and floating-point values may not decide an outcome ' +
        '(CONTRACTS section 4)',
    );
  }
  if (value < MIN_SAFE || value > MAX_SAFE) {
    throw new CanonicalizationError(
      `integer ${value} is outside the JSON safe-integer range; overflow must be rejected`,
    );
  }
  // String(-0) is "0", matching the Python implementation's normalization of -0.0.
  return String(value === 0 ? 0 : value);
}

function serialize(value: unknown): string {
  if (value === null) return 'null';
  if (value === true) return 'true';
  if (value === false) return 'false';
  if (typeof value === 'string') return serializeString(value);
  if (typeof value === 'number') return serializeNumber(value);
  if (typeof value === 'function' || typeof value === 'symbol' || value === undefined) {
    throw new CanonicalizationError(`${typeof value} has no JSON representation`);
  }
  if (typeof value === 'bigint') {
    throw new CanonicalizationError('bigint cannot be canonicalized within the safe-integer range');
  }
  if (Array.isArray(value)) {
    // Array.prototype.map skips holes, so a sparse array like [, 1] would serialize to "[,1]",
    // which is not valid JSON and has no Python counterpart.
    const parts: string[] = [];
    for (let i = 0; i < value.length; i += 1) {
      if (!(i in value)) {
        throw new CanonicalizationError(
          `array index ${i} is a hole; sparse arrays have no JSON representation`,
        );
      }
      parts.push(serialize(value[i]));
    }
    return `[${parts.join(',')}]`;
  }
  if (typeof value === 'object') {
    // Object.keys returns [] for a Date, Map, Set, RegExp or any class instance whose state is
    // not an own enumerable property, so these would serialize as "{}" and digest to a
    // confident, wrong hash. The Python implementation refuses them, so this must too — the same
    // silent-success failure mode the surrogate check exists to prevent.
    const prototype = Object.getPrototypeOf(value) as object | null;
    if (prototype !== Object.prototype && prototype !== null) {
      throw new CanonicalizationError(
        `${(value as object).constructor?.name ?? 'object'} has no JSON representation; ` +
          'only plain objects, arrays and JSON primitives can be canonicalized',
      );
    }
    const record = value as Record<string, unknown>;
    // Array.prototype.sort on strings compares UTF-16 code units, which is exactly what RFC8785
    // requires. The Python side has to ask for this explicitly.
    const keys = Object.keys(record);
    // Validate before sorting so both implementations refuse at the same point.
    for (const key of keys) rejectLoneSurrogates(key);
    keys.sort();
    const parts = keys.map((key) => {
      const entry = record[key];
      if (entry === undefined) {
        throw new CanonicalizationError(`property ${key} is undefined; JSON has no such value`);
      }
      return `${serializeString(key)}:${serialize(entry)}`;
    });
    return `{${parts.join(',')}}`;
  }
  throw new CanonicalizationError(`type ${typeof value} cannot be canonicalized`);
}

export function canonicalize(value: unknown): string {
  return serialize(value);
}

export function digest(value: unknown): string {
  return createHash('sha256').update(canonicalize(value), 'utf8').digest('hex');
}
